tests/test_models.py

```diff
@@ -0,0 +1,1001 @@
+    ]
+    with pytest.raises(ValueError, match="Images must have the same dimensions."):
+        _ = Patch_Dataset(img_list)
+
+    # list of ndarrays with HW and HWC mixed up
+    img_list = [
+        np.random.randint(0, 255, (4, 4, 3)),
+        np.random.randint(0, 255, (4, 4)),
+    ]
+    with pytest.raises(
+        ValueError, match="Each sample must be an array of the form HWC."
+    ):
+        _ = Patch_Dataset(img_list)
+
+    # list of mixed dtype
+    img_list = [np.random.randint(0, 255, (4, 4, 3)), "you_should_crash_here", 123, 456]
+    with pytest.raises(
+        ValueError,
+        match="Input must be either a list/array of images or a list of "
+        "valid image paths.",
+    ):
+        _ = Patch_Dataset(img_list)
+
+    # list of mixed dtype
+    img_list = ["you_should_crash_here", 123, 456]
+    with pytest.raises(
+        ValueError,
+        match="Input must be either a list/array of images or a list of "
+        "valid image paths.",
+    ):
+        _ = Patch_Dataset(img_list)
+
+    # list not exist paths
+    with pytest.raises(
+        ValueError,
+        match=r".*valid image paths.*",
+    ):
+        _ = Patch_Dataset(["img.npy"])
+
+    # ** test different extenstion parser
+    # save dummy data to temporary location
+    save_dir_path = os.path.join(rcParam["TIATOOLBOX_HOME"], "tmp_check/")
+    # remove prev generated data - just a test!
+    if os.path.exists(save_dir_path):
+        shutil.rmtree(save_dir_path, ignore_errors=True)
+    os.makedirs(save_dir_path)
+    torch.save({"a": "a"}, os.path.join(save_dir_path, "sample1.tar"))
+    np.save(
+        os.path.join(save_dir_path, "sample2.npy"), np.random.randint(0, 255, (4, 4, 3))
+    )
+
+    img_list = [
+        os.path.join(save_dir_path, "sample1.tar"),
+        os.path.join(save_dir_path, "sample2.npy"),
+    ]
+    with pytest.raises(
+        ValueError,
+        match=r"Can not load data of .*",
+    ):
+        _ = Patch_Dataset(img_list)
+    shutil.rmtree(rcParam["TIATOOLBOX_HOME"])
+
+    # preproc func for not defined dataset
+    with pytest.raises(
+        ValueError,
+        match=r".* preprocessing .* does not exist.",
+    ):
+        predefined_preproc_func("secret_dataset")
+
+
+def test_kather_patch_dataset():
+    """Test for kather patch dataset."""
+    size = (224, 224, 3)
+    # test kather with default param
+    dataset = Kather_Patch_Dataset()
+    # kather with default data path skip download
+    dataset = Kather_Patch_Dataset()
+    # pytest for not exist dir
+    with pytest.raises(
+        ValueError,
+        match=r".*not exist.*",
+    ):
+        _ = Kather_Patch_Dataset(save_dir_path="unknown_place")
+    # save to temporary location
+    save_dir_path = os.path.join(rcParam["TIATOOLBOX_HOME"], "tmp_check/")
+    # remove prev generated data - just a test!
+    if os.path.exists(save_dir_path):
+        shutil.rmtree(save_dir_path, ignore_errors=True)
+    url = (
```

Comment:
what if this link breaks? we should have a backup.
we can use try catch probably.

Reply:
which line in particular? Here we just remove the file if it already exists at the specified location and then download it. Seems pretty fail safe to me?

Reply:
A mirror might be a good idea for a backup (if the licence on kather patches is permissive we could host a backup copy). Failing that a helpful error message.

Reply:
I mean the zendo link.